File: 16. leetcode Hot100/[39]组合总和.java
//给定一个无重复元素的数组 candidates 和一个目标数 target ，找出 candidates 中所有可以使数字和为 target 的组合。 
//
// candidates 中的数字可以无限制重复被选取。 
//
// 说明： 
//
// 
// 所有数字（包括 target）都是正整数。 
// 解集不能包含重复的组合。 
// 
//
// 示例 1： 
//
// 输入：candidates = [2,3,6,7], target = 7,
//所求解集为：
//[
//  [7],
//  [2,2,3]
//]
// 
//
// 示例 2： 
//
// 输入：candidates = [2,3,5], target = 8,
//所求解集为：
//[
//  [2,2,2,2],
//  [2,3,3],
//  [3,5]
//] 
//
// 
//
// 提示： 
//
// 
// 1 <= candidates.length <= 30 
// 1 <= candidates[i] <= 200 
// candidate 中的每个元素都是独一无二的。 
// 1 <= target <= 500 
// 
// Related Topics 数组 回溯算法 
// 👍 1337 👎 0


//leetcode submit region begin(Prohibit modification and deletion)
class Solution {
    private void backTrace(int[] candidates, int target, int i, int boxSum, List<Integer> box, List<List<Integer>> ans) {
        final int N = candidates==null ?0: candidates.length;
        if (boxSum == target) {
            ans.add(new ArrayList<>());
            for (Integer x : box) {
                ans.get(ans.size()-1).add(x);
            }
        }
        if (boxSum>= target || i >= N) {
            return;
        }

        for (int j = i; j < N; j++) {
            box.add(candidates[j]);
            boxSum += candidates[j];
            backTrace(candidates, target, j, boxSum, box, ans);
            box.remove(box.size()-1);
            boxSum -= candidates[j];
        }
    }

    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        if (candidates == null || candidates.length == 0) {
            return new ArrayList<>();
        }
        List<Integer> box = new ArrayList<>();
        List<List<Integer>> ans = new ArrayList<>();
        int boxSum = 0;
        backTrace(candidates, target, 0, boxSum, box, ans);
        return ans;
    }
}
//leetcode submit region end(Prohibit modification and deletion)

File: 11. backTrace/[39]组合总和.py
# 给你一个 无重复元素 的整数数组 candidates 和一个目标整数 target ，找出 candidates 中可以使数字和为目标数 target 的
#  所有 不同组合 ，并以列表形式返回。你可以按 任意顺序 返回这些组合。 
# 
#  candidates 中的 同一个 数字可以 无限制重复被选取 。如果至少一个数字的被选数量不同，则两种组合是不同的。 
# 
#  对于给定的输入，保证和为 target 的不同组合数少于 150 个。 
# 
#  
# 
#  示例 1： 
# 
#  
# 输入：candidates = [2,3,6,7], target = 7
# 输出：[[2,2,3],[7]]
# 解释：
# 2 和 3 可以形成一组候选，2 + 2 + 3 = 7 。注意 2 可以使用多次。
# 7 也是一个候选， 7 = 7 。
# 仅有这两种组合。 
# 
#  示例 2： 
# 
#  
# 输入: candidates = [2,3,5], target = 8
# 输出: [[2,2,2,2],[2,3,3],[3,5]] 
# 
#  示例 3： 
# 
#  
# 输入: candidates = [2], target = 1
# 输出: []
#  
# 
#  
# 
#  提示： 
# 
#  
#  1 <= candidates.length <= 30 
#  2 <= candidates[i] <= 40 
#  candidates 的所有元素 互不相同 
#  1 <= target <= 40 
#  
# 
#  Related Topics 数组 回溯 👍 2668 👎 0
from typing import List

# leetcode submit region begin(Prohibit modification and deletion)
class Solution:
    def backtrace(self, nums, i, target, box, ans):
        N = 0 if not nums else len(nums)

        if sum(box) == target:
            ans.append([])
            for x in box:
                ans[-1].append(x)

        if sum(box) >=target or i>=N:
            return

        for j in range(i, N):
            box.append(nums[j])
            self.backtrace(nums, j, target, box, ans)
            box.pop()


    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        box = []
        ans = []
        self.backtrace(candidates, 0, target, box, ans)
        return ans

# leetcode submit region end(Prohibit modification and deletion)
